src/osd/PG.cc

```diff
@@ -605,20 +605,20 @@ void PG::merge_from(map<spg_t,PGRef>& sources, PeeringCtx &rctx,
   snap_mapper.update_bits(split_bits);
 }
 
-void PG::add_backoff(SessionRef s, const hobject_t& begin, const hobject_t& end)
+void PG::add_backoff(const ceph::ref_t<Session>& s, const hobject_t& begin, const hobject_t& end)
 {
-  ConnectionRef con = s->con;
+  auto& con = s->con;
```

Comment:
this cleanup changes the behavior and is buggy. we need to hold a strong reference of `s->con` here.

```
ceph version 15.0.0-4003-g130e2d2 (130e2d2f6c88df1c1be2ff3acacf9e46c0b72f88) octopus (dev)
 1: (()+0xf5d0) [0x7f0fc298d5d0]
 2: (PG::add_backoff(boost::intrusive_ptr<Session> const&, hobject_t const&, hobject_t const&)+0x3ea) [0x55c65b497faa]
 3: (PrimaryLogPG::do_request(boost::intrusive_ptr<OpRequest>&, ThreadPool::TPHandle&)+0xce5) [0x55c65b55d305]
 4: (OSD::dequeue_op(boost::intrusive_ptr<PG>, boost::intrusive_ptr<OpRequest>, ThreadPool::TPHandle&)+0x349) [0x55c65b3fdca9]
 5: (PGOpItem::run(OSD*, OSDShard*, boost::intrusive_ptr<PG>&, ThreadPool::TPHandle&)+0x69) [0x55c65b636059]
 6: (OSD::ShardedOpWQ::_process(unsigned int, ceph::heartbeat_handle_d*)+0x1427) [0x55c65b419fc7]
 7: (ShardedThreadPool::shardedthreadpool_worker(unsigned int)+0x5b6) [0x55c65b9cfe76]
 8: (ShardedThreadPool::WorkThreadSharded::entry()+0x10) [0x55c65b9d1fd0]
 9: (()+0x7dd5) [0x7f0fc2985dd5]
 10: (clone()+0x6d) [0x7f0fc184c02d]
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.
```

all of the crashes i checked in http://pulpito.ceph.com/kchai-2019-08-19_10:42:03-rados-wip-kefu-testing-2019-08-16-2103-distro-basic-smithi/ are caused by this.

Reply:
Thanks, I've fixed this and rebased.